kinder/ci/tools/update-workflows/templates/workflows/control-plane-local-kubelet-mode-tasks.yaml

```diff
@@ -0,0 +1,157 @@
+# IMPORTANT! this workflow is imported by control-plane-local-kubelet-mode-* workflows.
+version: 1
+summary: |
+  This workflow implements a sequence of tasks used test the proper functioning
+  of kubeadm not in control plane local kubelet mode during joining control plane.
+vars:
+  # vars defines default values for variable used by tasks in this workflow;
+  # those values might be overridden when importing this files.
+  kubernetesVersion: v1.31.0
+  upgradeVersion: v1.31.0
+  controlPlaneNodes: 3
+  workerNodes: 2
+  baseImage: kindest/base:v20221102-76f15095 # has containerd
+  image: kindest/node:test
+  clusterName: kinder-control-plane-local-kubelet-mode
+  kubeadmVerbosity: 6
+tasks:
+- name: pull-base-image
+  description: |
+    pulls kindest/base image with docker in docker and all the prerequisites necessary for running kind(er)
+  cmd: docker
+  args:
+  - pull
+  - "{{ .vars.baseImage }}"
+- name: add-kubernetes-versions
+  description: |
+    creates a node-image-variant by adding a Kubernetes version
+  cmd: kinder
+  args:
+  - build
+  - node-image-variant
+  - --base-image={{ .vars.baseImage }}
+  - --image={{ .vars.image }}
+  - --with-init-artifacts={{ .vars.kubernetesVersion }}
+  - --with-upgrade-artifacts={{ .vars.upgradeVersion }}
+  - --loglevel=debug
+  timeout: 15m
+- name: create-cluster
+  description: |
+    create a set of nodes ready for hosting the Kubernetes cluster
+  cmd: kinder
+  args:
+  - create
+  - cluster
+  - --name={{ .vars.clusterName }}
+  - --image={{ .vars.image }}
+  - --control-plane-nodes={{ .vars.controlPlaneNodes }}
+  - --worker-nodes={{ .vars.workerNodes }}
+  - --loglevel=debug
+  timeout: 5m
+- name: init
+  description: |
+    Initializes the Kubernetes cluster with version "initVersion"
+    by starting the boostrap control-plane nodes
+  cmd: kinder
+  args:
+  - do
+  - kubeadm-init
+  - --name={{ .vars.clusterName }}
+  - --loglevel=debug
+  - --kubeadm-verbosity={{ .vars.kubeadmVerbosity }}
+  - --kubeadm-feature-gate="ControlPlaneKubeletLocalMode=true"
+  - --copy-certs=auto
+  timeout: 5m
+- name: join
```

Comment:
You mean `post-join`, right? because the feature does not change the behaviour for control plane nodes which did `kubeadm init`?

Reply:
oh, so that's a problem i think if the feature is enabled we should make the `init` node kubelet.conf also point to local.
it would be weird for kubeadm users to have the feature enabled but one of the CP node kubelts will continue to point to the LB, while the rest point to local apiserver.

i know that for CAPI the init node is less important it's rolled out on upgrade.

Reply:
@chrischdi 

here is where it can be added for init:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/init/kubeletfinalize.go#L76
- two new phases has to be added
`kubelet-restart`, `kubelet-local-mode`

kubelet-restart is a new phase because with multiple phases that need restart we need just one at the end
both can be hidden and only called if the FG is enabled. the restart in kubelet-finalize can be skipped if the new phase kubelet-restart is called when the FG is enabled.
the new kubelet-local-node can be tagged EXPERIMENTAL and removed once the FG goes GA similarly to the phases in join. there could be some shared code between the kubelet-finalize and kubelet-local-mode as they both update the kubelet.conf file.

for upgrade:
- we can add a new function here https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L283 called something like `UpdateKubeletLocalMode`

for upgrade apply if the FG is enabled it can be called here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L177

for upgrade node and again if the FG is enabled it can be called here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/upgrade/node/controlplane.go#L86

the new function should respect the dry-run option. any fmt.Printf calls should be outside it IMO.

i explain it in detail but the changes to init/upgrade is not big.

this is something we missed discussing on the KEP. sorry about that..i naively assumed it as "standard practice" for new kubeadm features.

Reply:
No worries, I also already started to look at it :-)

Regarding init: Shouldn't it be enough to adjust the runKubeConfigFie function when run for the `kubelet` constant to write the local address?

Something like this:

```diff
diff --git a/cmd/kubeadm/app/cmd/phases/init/kubeconfig.go b/cmd/kubeadm/app/cmd/phases/init/kubeconfig.go
index 77432a13354..7e255b29fbb 100644
--- a/cmd/kubeadm/app/cmd/phases/init/kubeconfig.go
+++ b/cmd/kubeadm/app/cmd/phases/init/kubeconfig.go
@@ -26,6 +26,7 @@ import (
        "k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
        cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
        kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+       "k8s.io/kubernetes/cmd/kubeadm/app/features"
        kubeconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
        kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 )
@@ -157,7 +158,16 @@ func runKubeConfigFile(kubeConfigFileName string) func(workflow.RunData) error {
                cfg.CertificatesDir = data.CertificateWriteDir()
                defer func() { cfg.CertificatesDir = data.CertificateDir() }()

+               initConfiguration := data.Cfg().DeepCopy()
+
+               if features.Enabled(cfg.FeatureGates, features.ControlPlaneKubeletLocalMode) {
+                       if kubeConfigFileName == kubeadmconstants.KubeletKubeConfigFileName {
+                               // Unset the ControlPlaneEndpoint so the creation falls back to the LocalAPIEndpoint for the kubelet's kubeconfig.
+                               initConfiguration.ControlPlaneEndpoint = ""
+                       }
+               }
+
                // creates the KubeConfig file (or use existing)
-               return kubeconfigphase.CreateKubeConfigFile(kubeConfigFileName, data.KubeConfigDir(), data.Cfg())
+               return kubeconfigphase.CreateKubeConfigFile(kubeConfigFileName, data.KubeConfigDir(), initConfiguration)
        }
 }
```

Which then get's active when we run `NewKubeConfigFilePhase(kubeadmconstants.KubeletKubeConfigFileName)`?

Reply:
it seems hacky with the deepcopy but can work.
i guess adding extra phases for init is not needed.

Reply:
Opened https://github.com/kubernetes/kubernetes/pull/125780 for first feedback.

Will run tests on that now (trying to use the workflow implemented in this PR)

Reply:
> Opened [kubernetes/kubernetes#125780](https://github.com/kubernetes/kubernetes/pull/125780) for first feedback.
> 
> Will run tests on that now (trying to use the workflow implemented in this PR)

i will review a bit later